### PR TITLE
[Feature] add split context to connector framework

### PR DESCRIPTION
**Pull request by dirtysalt**
> Why I'm doing:

To support morsel dynamically in connector framework.

> What I'm doing:

Several things:
- add `ScanSplitContext` to `Morsel` struct
- connector implementation can implement split process by
   - defining how to split a scan range
   - how to use the splitted scan range
 - connector framework will provide necessary feature to support splitted scan range

Plus the PR includes some refactors
- remove unused methods an fields in `HdfsScanner` class
- simplify `scan_ranges` because we only have one scan range for each data source.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/connector/hive_connector.h [hunk 1]
```diff
@@ -77,6 +77,8 @@ class HiveDataSource final : public DataSource {
     int64_t estimated_mem_usage() const override;
     bool can_estimate_mem_usage() const override { return true; }
 
+    void get_split_tasks(std::vector<pipeline::ScanSplitContextPtr>* split_tasks) override;
```
Comment (tracymacding) [inline be/src/connector/hive_connector.h:80]: Only for hive connector, who will implement this api for other DataSource, LakeDataSource for example?

File: be/src/exec/hdfs_scanner.h [hunk 2]
```diff
@@ -311,6 +311,7 @@ class HdfsScanner {
     virtual Status do_init(RuntimeState* runtime_state, const HdfsScannerParams& scanner_params) = 0;
     virtual void do_update_counter(HdfsScanProfile* profile);
     virtual bool is_jni_scanner() { return false; }
+    virtual void get_split_tasks(std::vector<pipeline::ScanSplitContextPtr>* split_tasks) {}
```
Comment (tracymacding) [inline be/src/exec/hdfs_scanner.h:314]: To be implemented?

File: be/src/connector/hive_connector.h (hunk 1 above)
Comment (dirtysalt) [reply be/src/connector/hive_connector.h:80]: Anyone who cares can implement it.

File: be/src/exec/hdfs_scanner.h (hunk 2 above)
Comment (dirtysalt) [reply be/src/exec/hdfs_scanner.h:314]: Yes, to be implemented in next PR.

**Review by stephen-shelby — DISMISSED** [review]

**Review by stephen-shelby — DISMISSED** [review]

**Review by ZiheLiu — DISMISSED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by dirtysalt** [comment]
@mergify backport branch-3.2

**Pull request merged**